lib/matplotlib/backends/backend_gtk3.py

```diff
@@ -827,6 +848,8 @@ class _BackendGTK3(_Backend):
 
     @staticmethod
     def mainloop():
-        if Gtk.main_level() == 0:
-            cbook._setup_new_guiapp()
-            Gtk.main()
+        global _application
+
+        _application.run()  # Quits when all added windows close.
+        # Running after quit is undefined, so create a new one next time.
+        _application = None
```

Comment:
Somewhat of an edge case, but say someone sets up their own GtkApplication and this gets picked up by _create_application(), and then the application quits, and then we somehow reach again mainloop().  If I understand correctly, we need to set up a new GtkApplication in that case, because the old one will be invalid?  If Gtk provides no builtin way to know whether a GtkApplication has already been run, should _create_application e.g. connect to the `shutdown` signal to add a flag to the application when it shuts down?  Then the whole logic of whether new GtkApplications need to be created can be encapsulated in _create_application.

(Or I may have understood everything wrong...)

Reply:
Well, we add the windows to the `GtkApplication`, so if we create a new one here, it'd just exit immediately without windows attached. So they'd have to create a new window, and that would create a new application, I think? We could also exit early here if there's no application; is that needed if `plt.show` is called without any windows?

Reply:
I think(?) you missed my point (which upon reading again my message was somewhat unclear), which is mostly that you cannot tell a priori whether `_application` (if returned by Gio.Application.get_default()) is a fresh GtkApplication or an already used one (i.e. one on which run() has already been called and exited).  Or perhaps you can?

(Looking at it again, the point about `shutdown` was kind of a tangent.)

Reply:
No, I don't think you can tell that.

That's also why I only put `_create_application` in `FigureManager`, because I thought that'd only be called if we're managing the mainloop entirely. Checking for a default was a copy from Qt, but in this way, I think that it can't even happen?

Reply:
In that case (also re: https://github.com/matplotlib/matplotlib/pull/20284#issuecomment-851730262), does that mean this part (resetting _application to None so that a later call to _create_application() creates a new one) can/should be removed?

Reply:
If all windows are closed, then this (and thus `pyplot.show`) returns, so we still need to set it to `None` so we can get a new application for any subsequent windows. Otherwise, re-using an already run application will crash, as noted below.